### PR TITLE
Implement bid() and withdraw()

### LINKED ISSUE #15: Allow people to withdraw their funds once the auction is over or cancelled

### PR DESCRIPTION
**Pull request by mertcelebi**
Fixes #15 and #21

### CODE REVIEW

File: contracts/AuctionBase.sol [hunk 1]
```diff
@@ -129,59 +128,101 @@ contract AuctionBase is Pausable {
...
     BidCreated(_auctionId, auction.nftAddress, auction.tokenId, msg.sender, newBid);
+    return true;
+  }
+
+  // @dev Allow people to withdraw their balances or the NFT
+  function withdrawBalance(uint256 _auctionId) external returns (bool success) {
+    AuctionStatus _status = _getAuctionStatus(_auctionId);
+
+    // Don't let people withdraw before the auction is cancelled or completed
+    require(_status == AuctionStatus.Cancelled || _status == AuctionStatus.Completed);
```
Comment (mykelp) [inline contracts/AuctionBase.sol]: Someone should be able to withdraw their funds if not the highest bidder, while auction is still accepting bids

File: contracts/AuctionBase.sol [hunk 2]
```diff
@@ -129,59 +128,101 @@ contract AuctionBase is Pausable {
...
+
+    Auction storage auction = auctions[_auctionId];
+    address withdrawalAccount;
```
Comment (mykelp) [inline contracts/AuctionBase.sol]: Shouldn't withdrawalAccount always be `msg.sender`?

File: contracts/AuctionBase.sol [hunk 3]
```diff
@@ -129,59 +128,101 @@ contract AuctionBase is Pausable {
...
+      // Anyone else gets what they bid
+      else {
+        withdrawalAccount = msg.sender;
+        withdrawalAmount = auction.fundsByBidder[withdrawalAccount];
+      }
+    }
+
+    require(withdrawalAmount > 0);
+    auction.fundsByBidder[withdrawalAccount] -= withdrawalAmount;
+    msg.sender.transfer(withdrawalAmount);
```
Comment (mertcelebi) [inline contracts/AuctionBase.sol]: This is wrong, needs fixing.

File: contracts/AuctionBase.sol (hunk 2 above)
Comment (mertcelebi) [reply contracts/AuctionBase.sol]: Nope, who is receiving the money is the `msg.sender`. `withdrawalAccount` is basically the source of the money. And `msg.sender` is the withdrawer. How can we name it better?

File: contracts/AuctionBase.sol (hunk 1 above)
Comment (mertcelebi) [reply contracts/AuctionBase.sol]: Good call

Comment (mertcelebi) [reply contracts/AuctionBase.sol]: Could make the logic a little weird, but will look into it

File: contracts/AuctionBase.sol (hunk 2 above)
Comment (mertcelebi) [reply contracts/AuctionBase.sol]: 👍

**Pull request merged**